sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -691,32 +731,115 @@ ssize_t coap_opt_finish(coap_pkt_t *pkt, uint16_t flags)
     return pkt->payload - (uint8_t *)pkt->hdr;
 }
 
+void coap_block2_init(coap_pkt_t *pkt, coap_block_slicer_t *slicer)
+{
+    uint32_t blknum;
+    unsigned szx;
+    /* Retrieve the block2 option from the client request */
+    if(coap_get_blockopt(pkt, COAP_OPT_BLOCK2, &blknum, &szx) >= 0) {
+        /* Use the client requested block size if it is smaller than our own
+         * maximum block size */
+        if (NANOCOAP_BLOCK_SIZE_EXP_MAX - 4 < szx ) {
+            szx =NANOCOAP_BLOCK_SIZE_EXP_MAX - 4;
+        }
+    }
+    slicer->start = blknum * coap_szx2size(szx);
+    slicer->end = slicer->start + coap_szx2size(szx);
+    slicer->cur = 0;
+}
+
+void coap_block2_finish(coap_pkt_t *pkt, coap_block_slicer_t *slicer)
+{
+    int option_len;
+    uint16_t delta;
+    /* Retrieve the block2 option from the constructed packet */
+    _parse_option(pkt, slicer->opt, &delta, &option_len);
```

Comment:
Use of pkt is not valid here. The pkt referred to here is the _response_ packet, not the request packet as in coap_block2_init(). This struct has not been initialized for the response. In fact, when using the buffer-based functions for the response, the pkt never is updated for the response.

All we really need to determine is the delta value. We can directly use _decode_value(). The only issue is the third parameter, which points to the beginning of the payload to be sure the function does not read past the end of the header. We *know* it won't read past the end of the header because the option already was written before getting here. So, we can just point to the location past the longest possible value for a block2 option. 

That's what I did in [81a818b](https://github.com/bergzand/RIOT/pull/3/commits/81a818b55dff53267c087a12bf2b1e917e71b53d).